File: 72. Edit Distance/72. Edit Distance.rb
# @param {String} word1
# @param {String} word2
# @return {Integer}
def min_distance(word1, word2)
    m, n = word1.length, word2.length
    dp = Array.new(m + 1) { Array.new(n + 1, 0) }
    
    (1..m).each { |i| dp[i][0] = i }
    (1..n).each { |j| dp[0][j] = j }
    
    (1..m).each do |i|
        (1..n).each do |j|
            if word1[i-1] == word2[j-1]
                dp[i][j] = dp[i-1][j-1]
            else
                dp[i][j] = [dp[i-1][j], dp[i][j-1], dp[i-1][j-1]].min + 1
            end
        end
    end
    
    dp[m][n]
end

File: 72. Edit Distance/72. Edit Distance.js
function minDistance(word1, word2) {
    const m = word1.length;
    const n = word2.length;

    // Create a DP table
    const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));

    // Initialize the DP table
    for (let i = 0; i <= m; i++) {
        dp[i][0] = i;
    }
    for (let j = 0; j <= n; j++) {
        dp[0][j] = j;
    }

    // Fill the DP table
    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (word1[i - 1] === word2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1];  // No additional cost if characters are the same
            } else {
                dp[i][j] = Math.min(
                    dp[i - 1][j] + 1,    // Deletion
                    dp[i][j - 1] + 1,    // Insertion
                    dp[i - 1][j - 1] + 1 // Replacement
                );
            }
        }
    }

    return dp[m][n];
}

// Example usage:
console.log(minDistance("horse", "ros")); // Output: 3
console.log(minDistance("intention", "execution")); // Output: 5